pkg/build/strategies/sti/sti.go

```diff
@@ -0,0 +1,486 @@
+package sti
+
+import (
+	"bufio"
+	"bytes"
+	"fmt"
+	"io"
+	"path/filepath"
+	"regexp"
+	"time"
+
+	"github.com/golang/glog"
+	"github.com/openshift/source-to-image/pkg/api"
+	"github.com/openshift/source-to-image/pkg/build"
+	"github.com/openshift/source-to-image/pkg/docker"
+	"github.com/openshift/source-to-image/pkg/errors"
+	"github.com/openshift/source-to-image/pkg/git"
+	"github.com/openshift/source-to-image/pkg/script"
+	"github.com/openshift/source-to-image/pkg/tar"
+	"github.com/openshift/source-to-image/pkg/util"
+)
+
+const (
+	// maxErrorOutput is the maximum length of the error output saved for processing
+	maxErrorOutput = 1024
+	// defaultLocation is the default location of the scripts and sources in image
+	defaultLocation = "/tmp"
+)
+
+var (
+	// List of directories that needs to be present inside workign dir
+	workingDirs = []string{
+		"upload/scripts",
+		"upload/src",
+		"downloads/scripts",
+		"downloads/defaultScripts",
+	}
+)
+
+// STI strategy executes the STI build.
+// For more details about STI, visit https://github.com/openshift/source-to-image
+type STI struct {
+	request         *api.Request
+	result          *api.Result
+	postExecutor    docker.PostExecutor
+	installer       script.Installer
+	git             git.Git
+	fs              util.FileSystem
+	tar             tar.Tar
+	docker          docker.Docker
+	callbackInvoker util.CallbackInvoker
+	requiredScripts []api.Script
+	optionalScripts []api.Script
+
+	// Interfaces
+	builder     build.LayeredDockerBuilder
+	preparer    build.Preparer
+	incremental build.IncrementalBuilder
+	scripts     build.ScriptsHandler
+	source      build.Downloader
+	garbage     build.Cleaner
+}
+
+// New returns a new STI builder
+func New(req *api.Request) (*STI, error) {
+	docker, err := docker.New(req.DockerSocket)
+	if err != nil {
+		return nil, err
+	}
+	inst := script.NewInstaller(req.BaseImage, req.ScriptsURL, req.InstallDestination, docker)
+
+	b := STI{
+		installer:       inst,
+		request:         req,
+		docker:          docker,
+		git:             git.New(),
+		fs:              util.NewFileSystem(),
+		tar:             tar.New(),
+		callbackInvoker: util.NewCallbackInvoker(),
+		requiredScripts: []api.Script{api.Assemble, api.Run},
+		optionalScripts: []api.Script{api.SaveArtifacts},
+	}
+
+	// The sources are downloaded using the GIT downloader.
+	// TODO: Add more SCM in future.
+	b.source = &git.Clone{b.git, b.fs}
+
+	b.garbage = &build.DefaultCleaner{b.fs, b.docker}
+
+	// Set interfaces
+	b.preparer = &b
```

Comment:
Why expose these interfaces like this?  So you can test it?  The fact that you are keeping references to yourself to do this indicates there's a missing refactor here (which you don't have to solve now, but I want to point it out so at least it's in your mind).  

Also, you can nuke about ten of the & by using using `&STI` up above.